src/Builders/PromptBuilder.php

```diff
@@ -1040,67 +1109,179 @@ private function getConfiguredModel(CapabilityEnum $capability): ModelInterface
-                        implode(', ', array_map(function ($opt) {
-                            return $opt->getName()->value . '=' . json_encode($opt->getValue());
-                        }, $requirements->getRequiredOptions()))
-                    )
-                );
+            if (!empty($matchingPreferences)) {
+                // Get the first matching preference key
+                $firstMatchKey = key($matchingPreferences);
+                [$providerId, $modelId] = $candidateMap[$firstMatchKey];
+
+                $selectedModel = $this->registry->getProviderModel($providerId, $modelId, $this->modelConfig);
             }
+        }
 
-            $provider = $this->providerIdOrClassName;
-            $modelMetadata = $modelsMetadata[0];
+        if ($selectedModel === null) {
+            // No preference matched; fall back to the first candidate discovered.
+            [$providerId, $modelId] = reset($candidateMap);
+
+            $selectedModel = $this->registry->getProviderModel(
+                $providerId,
+                $modelId,
+                $this->modelConfig
+            );
         }
 
-        // Get the model instance from the provider
-        return $this->registry->getProviderModel(
-            $provider,
-            $modelMetadata->getId(),
-            $this->modelConfig
+        return $selectedModel;
+    }
+
+    /**
+     * Builds a map of candidate models that satisfy the requirements for efficient lookup.
+     *
+     * @since n.e.x.t
+     *
+     * @param ModelRequirements $requirements The requirements derived from the prompt.
+     * @return array<string, array{0:string,1:string}> Map of preference keys to [providerId, modelId] tuples.
+     */
+    private function getCandidateModelsMap(ModelRequirements $requirements): array
+    {
+        if ($this->providerIdOrClassName === null) {
+            // No provider locked in, gather all models across providers that meet requirements.
+            $providerModelsMetadata = $this->registry->findModelsMetadataForSupport($requirements);
+
+            $candidateMap = [];
+            foreach ($providerModelsMetadata as $providerModels) {
+                $providerId = $providerModels->getProvider()->getId();
+                $providerMap = $this->generateMapFromCandidates($providerId, $providerModels->getModels());
+                // Use + operator to merge, preserving keys from $candidateMap (first provider wins for model-only keys)
+                $candidateMap = $candidateMap + $providerMap;
+            }
+
+            return $candidateMap;
+        }
+
+        // Provider set, only consider models from that provider.
+        $modelsMetadata = $this->registry->findProviderModelsMetadataForSupport(
+            $this->providerIdOrClassName,
+            $requirements
         );
+
+        return $this->generateMapFromCandidates($this->providerIdOrClassName, $modelsMetadata);
```

Comment:
See my comment in https://github.com/WordPress/php-ai-client/pull/110#discussion_r2433846822, I think this is still not right here.